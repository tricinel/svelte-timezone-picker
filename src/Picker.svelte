<script>
  /* global __USE_CUSTOM_EVENT__ */
  import { createEventDispatcher, onMount } from 'svelte';
  import { get_current_component } from 'svelte/internal'; // eslint-disable-line camelcase
  import { slide } from 'svelte/transition';
  import groupedZones from './timezones';
  import {
    scrollIntoView,
    uid,
    slugify,
    keyCodes,
    ungroup,
    filter,
    pick
  } from './utils';

  // ***** Public API *****

  // The timezone value comes from the consumer of the component
  // If it's not provided, we will set it in onMount to be the user's current timezone
  export let timezone = null;

  // Should the dropdown be expanded by default?
  export let expanded = false;

  // We can allow the user to filter the timezones displayed to only a few
  export let allowedTimezones = null;

  // ***** End Public API *****

  // What is the current zone?
  // Array ['Abidjan', '+00:00', '+00:00']
  // The first value is the display name for the zone, the second is the standard offset, the third the daylight saving time offset
  let currentZone;

  // We keep track of what the user is typing in the search box
  // String
  let userSearch;

  // What is the currently selected zone in the dropdown?
  // String 'Africa/Abidjan'
  let highlightedZone;

  // DOM nodes refs
  let toggleButtonRef;
  let searchInputRef;
  let clearButtonRef;
  let listBoxRef;
  let listBoxOptionRefs;

  // A few IDs that will we use for a11y
  const labelId = uid();
  const listBoxId = uid();
  const searchInputId = uid();

  // We ungroup the zones
  // e.g. { Africa: {'Africa/Abidjan': ['Abidjan', '+00:00', '+00:00']} }
  // => {'Africa/Abidjan': ['Abidjan', '+00:00', '+00:00']}
  const ungroupedZones = ungroup(groupedZones);

  const userTimezone = Intl.DateTimeFormat().resolvedOptions().timeZone; // eslint-disable-line new-cap

  // We will only display the timezones the user passed in
  // and default to all the zones if that's empty or the wrong format
  let availableZones = ungroupedZones;

  if (allowedTimezones) {
    if (Array.isArray(allowedTimezones)) {
      availableZones = pick(ungroupedZones, [
        ...allowedTimezones,
        userTimezone
      ]);
    } else {
      console.error(
        'You need to provide a list of timezones as an Array!',
        `You provided ${allowedTimezones}.`
      );
    }
  }

  // We also want a list of all the valid zones
  // e.g. {'Africa/Abidjan': ['Abidjan', '+00:00', '+00:00'], 'Africa/Accra': ['Accra', '+00:00', '+00:00']}
  // => ['Africa/Abidjan', 'Africa/Accra']
  const validZones = Object.keys(availableZones);

  // Zones will be filtered as the user types, so we keep track of them internally here
  let filteredZones = [];

  // We take the ungroupedZones and create a list of just the user-visible labels
  // and add them to the refs
  // e.g. {'Africa/Abidjan': ['Abidjan', '+00:00', '+00:00'], 'Africa/Accra': ['Accra', '+00:00', '+00:00']}
  // => ['Abidjan', 'Accra']
  listBoxOptionRefs = Object.values(availableZones).map(([zone]) => ({ [zone]: null }));

  // We keep track of the initial state so we can reset to these values when needed
  const initialState = {
    expanded,
    userSearch: null
  };

  // Reset the dropdown and all internal state to the initial values
  const reset = () => {
    expanded = initialState.expanded; // eslint-disable-line prefer-destructuring
    userSearch = initialState.userSearch; // eslint-disable-line prefer-destructuring
  };

  // We will use the dispatcher to send the update event
  const dispatch = createEventDispatcher();
  // Because CustomEvents don't bubble by default, custom components won't work
  // We will need to do some tricks for this to work properly
  // https://github.com/sveltejs/svelte/issues/3119
  const component = get_current_component();

  const dispatchUpdates = () => {
    const eventName = 'update';
    const eventData = { timezone };

    // __USE_CUSTOM_EVENT__ is inserted through the rollup build process
    if (__USE_CUSTOM_EVENT__) {
      const customEvent = new CustomEvent(eventName, {
        detail: eventData,
        bubbles: true,
        cancelable: true,
        composed: true
      });
      component.dispatchEvent && component.dispatchEvent(customEvent);
    }

    if (!__USE_CUSTOM_EVENT__) {
      dispatch(eventName, eventData);
    }
  };

  // Emit the event back to the consumer
  const handleTimezoneUpdate = (ev, zoneId) => {
    currentZone = ungroupedZones[zoneId];
    timezone = zoneId;
    dispatchUpdates();
    reset();
    toggleButtonRef.focus();
    ev.preventDefault();
  };

  // ***** Methods *****

  // Figure out if a grouped zone has any currently visible zones
  // We use this when the user searches in order to show/hide the group name in the list
  const groupHasVisibleChildren = (group, zones) =>
    Object.keys(groupedZones[group]).some((zone) => zones.includes(zone));

  // Scroll the list to a specific element in it if that element is not already visible on screen
  const scrollList = (zone) => {
    const zoneElementRef = listBoxOptionRefs[zone];
    if (listBoxRef && zoneElementRef) {
      scrollIntoView(zoneElementRef, listBoxRef);
      zoneElementRef.focus({ preventScroll: true });
    }
  };

  // Every time the user uses their keyboard to move up or down in the list,
  // we need to figure out if their at the end/start of the list and scroll the correct elements
  // into view
  const moveSelection = (direction) => {
    const len = filteredZones.length;
    const zoneIndex = filteredZones.findIndex(
      (zone) => zone === highlightedZone
    );

    let index;

    if (direction === 'up') {
      index = (zoneIndex - 1 + len) % len;
    }

    if (direction === 'down') {
      index = (zoneIndex + 1) % len;
    }

    // We update the highlightedZone to be the one the user is currently on
    highlightedZone = filteredZones[index];
    // We make sure the highlightedZone is visible on screen, scrolling it into view if not
    scrollList(highlightedZone);
  };

  // We watch for when the user presses Escape, ArrowDown or ArrowUp and react accordingly
  const keyDown = (ev) => {
    // If the clearButton is focused, don't do anything else
    // We should only continue if the dropdown is expanded
    if (document.activeElement === clearButtonRef || !expanded) {
      return;
    }

    // If the user presses Escape, we dismiss the drodpown
    if (ev.keyCode === keyCodes.Escape) {
      reset();
    }

    // If the user presses the down arrow, start navigating the list
    if (ev.keyCode === keyCodes.ArrowDown) {
      ev.preventDefault();
      moveSelection('down');
    }
    // If the user presses the up arrow, start navigating the list
    if (ev.keyCode === keyCodes.ArrowUp) {
      ev.preventDefault();
      moveSelection('up');
    }
    // If the user presses Enter and the dropdown is expanded, select the current item
    if (ev.keyCode === keyCodes.Enter && highlightedZone) {
      handleTimezoneUpdate(ev, highlightedZone);
    }
    // If the user start to type letters or numbers, we focus on the Search field
    if (
      keyCodes.Characters.includes(ev.keyCode) ||
      ev.keyCode === keyCodes.Backspace
    ) {
      searchInputRef.focus();
    }
  };

  // When the user presses the clear button when searching,
  // we want to clear the text and refocus on the input
  const clearSearch = () => {
    userSearch = initialState.userSearch; // eslint-disable-line prefer-destructuring
    // Refocus to the search input
    searchInputRef.focus();
  };

  const setHighlightedZone = (zone) => {
    highlightedZone = zone;
  };

  const toggleExpanded = (ev) => {
    if (ev.keyCode) {
      // If it's a keyboard event, we should react only to certain keys
      // Enter and Space should show it
      if ([keyCodes.Enter, keyCodes.Space].includes(ev.keyCode)) {
        expanded = !expanded;
      }
      // Escape should just hide the menu
      if (ev.keyCode === keyCodes.Escape) {
        expanded = false;
      }
      // ArrowDown should show it
      if (ev.keyCode === keyCodes.ArrowDown) {
        expanded = true;
      }
    } else {
      // If there is no keyCode, it's not a keyboard event
      expanded = !expanded;
    }
  };

  const scrollToHighlighted = () => {
    if (expanded && highlightedZone) {
      scrollList(highlightedZone);
    }
  };

  // ***** Reactive *****

  // As the user types, we filter the available zones to show only those that should be visible
  $: filteredZones =
      userSearch && userSearch.length > 0
        ? filter(userSearch, availableZones)
        : validZones.slice();

  const setTimezone = (tz) => {
    if (!tz) {
      timezone = userTimezone;
    }

    if (tz && !validZones.includes(tz)) {
      // The timezone must be a valid timezone, so we check it against our list of values in flat
      console.warn(
        `The timezone provided is not valid: ${tz}!`,
        `Valid zones are: ${validZones}`
      );
      timezone = userTimezone;
    }

    currentZone = ungroupedZones[timezone];
    setHighlightedZone(timezone);
  };

  // We want to properly handle any potential changes to the current timezone
  // that might come in from the consumer of the component.
  // This includes setting the proper timezone and dispatching the updated values
  // back up to the consumer
  $: setTimezone(timezone);

  // ***** Lifecycle methods *****
  onMount(() => {
    setTimezone(timezone);
    scrollToHighlighted();
  });
</script>

{#if expanded}
  <div class="overlay" on:click="{reset}"></div>
{/if}

<button
  bind:this="{toggleButtonRef}"
  type="button"
  aria-label="{`${currentZone[0]} is currently selected. Change timezone`}"
  aria-haspopup="listbox"
  data-toggle="true"
  aria-expanded="{expanded}"
  on:click="{toggleExpanded}"
  on:keydown="{toggleExpanded}"
>
  <span>{currentZone[0]} <small>GMT {currentZone[1]}</small></span>
  <svg width="10" height="16" viewBox="0 0 16 16">
    <polygon x="0" y="0" points="8, 8, 16, 16, 0, 16" transform="{expanded ? 'rotate(0)' : 'rotate(180, 8, 8)'} translate(0 -4)"/>
  </svg>
</button>
{#if expanded}
  <div
    class="tz-dropdown"
    transition:slide
    on:introend="{scrollToHighlighted}"
    on:keydown="{keyDown}"
  >
    <span class="sr-only" id="{labelId}">
      Select a timezone from the list. Start typing to filter or use the arrow
      keys to navigate the list
    </span>
    <div class="input-group">
      <!-- svelte-ignore a11y-autofocus -->
      <input
        id="{searchInputId}"
        bind:this="{searchInputRef}"
        type="search"
        aria-autocomplete="list"
        aria-controls="{listBoxId}"
        aria-labelledby="{labelId}"
        autocomplete="off"
        autocorrect="off"
        placeholder="Search..."
        bind:value="{userSearch}"
        autofocus
      />

      {#if userSearch && userSearch.length > 0}
        <button
          bind:this="{clearButtonRef}"
          title="Clear search text"
          on:click="{clearSearch}"
        >
          &times;
        </button>
      {/if}
    </div>

    <ul
      tabindex="-1"
      class="tz-groups"
      id="{listBoxId}"
      role="listbox"
      bind:this="{listBoxRef}"
      aria-labelledby="{labelId}"
      aria-activedescendant="{currentZone && `tz-${slugify(currentZone[0])}`}"
    >
      {#each Object.keys(groupedZones) as group}
        {#if groupHasVisibleChildren(group, filteredZones)}
          <li role="option" aria-hidden="true">
            <p>{group}</p>
          </li>
          {#each Object.entries(groupedZones[group]) as [zoneLabel, zoneDetails]}
            {#if filteredZones.includes(zoneLabel)}
              <li
                role="option"
                tabindex="0"
                id="{`tz-${slugify(zoneLabel)}`}"
                bind:this="{listBoxOptionRefs[zoneLabel]}"
                aria-label="{`Select ${zoneDetails[0]}`}"
                aria-selected="{highlightedZone === zoneDetails[0]}"
                on:mouseover="{() => setHighlightedZone(zoneDetails[0])}"
                on:click="{(ev) => handleTimezoneUpdate(ev, zoneLabel)}"
              >
                {zoneDetails[0]} <span>GMT {zoneDetails[1]}</span>
              </li>
            {/if}
          {/each}
        {/if}
      {/each}
    </ul>
  </div>
{/if}

<style>
  .overlay {
    background: transparent;
    height: 100vh;
    left: 0;
    position: fixed;
    top: 0;
    width: 100vw;
    z-index: 0;
  }

  button {
    background: transparent;
    border: 0;
    cursor: pointer;
  }

  svg polygon {
    fill: var(--color-info-900, #076196);
  }

  button[data-toggle] {
    align-content: flex-start;
    align-items: center;
    display: flex;
    padding: 0;
  }

  button[data-toggle] > span {
    color: var(--color-info-900, #076196);
    font-weight: 500;
    margin-right: 0.4em;
    text-decoration: underline;
  }

  button[data-toggle] > span small {
    font-weight: 400;
    font-size: 0.8em;
  }

  .tz-dropdown {
    background-color: var(--color-white, #fff);
    border: 1px solid var(--color-gray-100, rgba(0, 0, 0, 0.2));
    box-shadow: 0 1px 6px 0 var(--color-gray-100, rgba(0, 0, 0, 0.2));
    border-radius: 4px;
    display: flex;
    flex-direction: column;
    min-width: 18em;
    max-width: 100vw;
    position: absolute;
    z-index: 50;
  }

  .tz-groups {
    height: 240px;
    max-height: 40vh;
    overflow: scroll;
  }

  ul {
    margin: 0;
    list-style: none inside none;
    padding: 0;
  }

  ul li {
    font-size: 0.9rem;
    display: block;
    margin: 0;
    padding: 0;
  }

  ul li > span {
    font-size: 0.8em;
    line-height: 1.4em;
    text-align: right;
  }

  ul li p {
    color: var(--color-gray-900, #292929);
    font-size: 0.92rem;
    font-weight: 600;
    letter-spacing: 0.08em;
    margin: 0;
    padding: 0;
    text-transform: uppercase;
  }

  ul li {
    background: transparent;
    border: 0;
    color: var(--color-gray-600, #757575);
    display: flex;
    justify-content: space-between;
    padding: 0.8em 1.2em;
    text-align: left;
  }

  ul li[aria-selected]:hover,
  ul li:focus,
  li[aria-selected='true'] {
    background: var(--color-info-900, #076196);
    color: #fff;
    cursor: pointer;
  }

  .input-group {
    display: flex;
    position: relative;
  }

  .input-group > button {
    position: absolute;
    top: 1.1em;
    right: 0.8em;
  }

  input {
    border: 1px solid var(--color-gray-400, #acacac);
    border-radius: 4px;
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.12);
    margin: 0.4em;
    padding: 0.6em 1.6em 0.6em 0.4em;
    width: calc(100% - 0.8em);
  }

  .sr-only {
    border: 0;
    clip: 'rect(0, 0, 0, 0)';
    height: 1px;
    margin: -1px;
    opacity: 0;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }
</style>
