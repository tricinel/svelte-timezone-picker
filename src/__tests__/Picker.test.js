import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent } from '@testing-library/svelte';
import userEvent from '@testing-library/user-event';

import {
  getZoneLabelAtIndex,
  getTestRegex,
  getTotalZones,
  filterZones,
  keyArrowDown,
  keyArrowUp,
  keyEnter,
  keyEscape
} from '../testUtils';
import Picker from '../Picker.svelte';

jest.mock('../timezones');

const props = {
  datetime: '2016-06-19T08:30',
  timezone: 'Europe/London'
};

describe('The component renders with internal defaults', () => {
  let intlSpy;

  beforeEach(() => {
    intlSpy = jest.spyOn(Intl, 'DateTimeFormat');
  });

  afterEach(() => {
    intlSpy.mockRestore();
    jest.clearAllTimers();
  });

  test(`Shows the browser's timezone selected when rendered without a timezone prop`, () => {
    intlSpy.mockReturnValueOnce({
      resolvedOptions: () => ({
        timeZone: 'UTC'
      })
    });

    const { getByText } = render(Picker, {
      datetime: '2016-06-19T08:30'
    });

    expect(getByText('UTC Time')).toBeInTheDocument();
  });

  test('Sets the datetime to the current datetime when rendered without a datetime prop', () => {
    const { getByText } = render(Picker, {
      timezone: 'Europe/Berlin'
    });

    expect(getByText('Central European Time')).toBeInTheDocument();
  });
});

describe('The component renders with props', () => {
  test('Shows the selected option when rendered with all props', () => {
    const { getByText, getByLabelText } = render(Picker, props);
    const toggleButton = getByLabelText(/Change timezone/i);

    expect(toggleButton).toHaveAttribute('aria-expanded', 'false');
    expect(getByText('London')).toBeInTheDocument();
  });

  test('Shows the listbox by default when the open prop is true', () => {
    const { getByLabelText, getByRole } = render(Picker, {
      ...props,
      open: true
    });
    const toggleButton = getByLabelText(/Change timezone/i);
    const listBox = getByRole('listbox');

    expect(toggleButton).toHaveAttribute('aria-expanded', 'true');
    expect(listBox).toBeInTheDocument();
  });
});

describe('The component handles user interactions', () => {
  test('The user can click or press Enter on the currently selected timezone and toggle the selector dropdown', async () => {
    const {
      queryByPlaceholderText,
      getByPlaceholderText,
      getByLabelText,
      getByRole,
      container
    } = render(Picker, props);
    const toggleButton = getByLabelText(/Change timezone/i);

    // Initially, the search field is not in the document
    expect(queryByPlaceholderText(/search/i)).not.toBeInTheDocument();
    expect(toggleButton).toHaveAttribute('aria-expanded', 'false');

    await fireEvent.click(toggleButton);
    expect(toggleButton).toHaveAttribute('aria-expanded', 'true');

    // After clicking the toggle button, the search field should be in the document
    expect(getByPlaceholderText(/search/i)).toBeInTheDocument();
    // So should the listbox
    expect(getByRole('listbox')).toBeInTheDocument();

    // We can now hide it
    await fireEvent.click(toggleButton);
    expect(toggleButton).toHaveAttribute('aria-expanded', 'false');

    // And we can show toggle expansion again using Enter
    toggleButton.focus();
    await fireEvent.keyUp(document.activeElement || document.body, keyEnter);

    // And hide it with Escape
    await fireEvent.keyUp(document.activeElement || document.body, keyEscape);
    expect(toggleButton).toHaveAttribute('aria-expanded', 'false');

    // And we can show toggle expansion again using Enter
    toggleButton.focus();
    await fireEvent.keyUp(document.activeElement || document.body, keyEnter);

    // And hide it by clicking outside of the dropdown, anywhere on the overlay
    const overlay = container.querySelector('.overlay');
    await fireEvent.click(overlay);
    expect(toggleButton).toHaveAttribute('aria-expanded', 'false');
  });

  test('The user can navigate the options using the keyboard', async () => {
    const { getByText, getByLabelText } = render(Picker, {
      ...props,
      open: true
    });

    const first = getZoneLabelAtIndex(0);
    const last = getZoneLabelAtIndex(getTotalZones() - 1);

    // The user pressed the down arrow key, so we move focus to the first option
    await fireEvent.keyUp(
      document.activeElement || document.body,
      keyArrowDown
    );
    expect(getByText(getTestRegex(first))).toHaveFocus();

    // The user pressed the down key, so we move focus to the second option
    await fireEvent.keyUp(
      document.activeElement || document.body,
      keyArrowDown
    );
    expect(getByText(getTestRegex(getZoneLabelAtIndex(1)))).toHaveFocus();

    // The user pressed the up key, so we move focus to the first option
    await fireEvent.keyUp(document.activeElement || document.body, keyArrowUp);
    expect(getByText(getTestRegex(first))).toHaveFocus();

    // The user pressed the up key, so we move focus to the last option because we are at the top of the list
    await fireEvent.keyUp(document.activeElement || document.body, keyArrowUp);
    expect(getByText(getTestRegex(last))).toHaveFocus();

    // The user pressed the Enter key, so we select the option currently focused
    // `last` after the above events
    await fireEvent.keyUp(document.activeElement || document.body, keyEnter);
    expect(
      getByLabelText(`${last} is currently selected. Change timezone`)
    ).toBeInTheDocument();
  });

  test('The use can type to filter the options', async () => {
    const { getAllByRole, getByPlaceholderText } = render(Picker, {
      ...props,
      open: true
    });
    const input = getByPlaceholderText(/search/i);

    await userEvent.type(input, 'bo');
    expect(getAllByRole('option')).toHaveLength(filterZones('bo').length);

    await userEvent.type(input, '{backspace}');
    expect(getAllByRole('option')).toHaveLength(filterZones('b').length);
  });
});
