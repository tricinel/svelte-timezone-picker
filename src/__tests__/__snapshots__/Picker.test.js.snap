// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component renders with internal defaults Basic snapshot with defaults 1`] = `
<div>
   
  <button
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-label="London is currently selected. Change timezone"
    class="svelte-1mifefb"
    data-toggle="true"
    type="button"
  >
    <span
      class="svelte-1mifefb"
    >
      London
       
      <small
        class="svelte-1mifefb"
      >
        GMT 
        +00:00
      </small>
    </span>
     
    <svg
      class="svelte-1mifefb"
      height="16"
      viewBox="0 0 16 16"
      width="10"
    >
      <polygon
        class="svelte-1mifefb"
        points="8, 8, 16, 16, 0, 16"
        transform="rotate(180, 8, 8) translate(0 -4)"
        x="0"
        y="0"
      />
    </svg>
  </button>
   
  
</div>
`;

exports[`The component renders with internal defaults Basic snapshot with defaults 2`] = `
<div>
  <div
    class="overlay svelte-1mifefb"
  />
   
  <button
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-label="London is currently selected. Change timezone"
    class="svelte-1mifefb"
    data-toggle="true"
    type="button"
  >
    <span
      class="svelte-1mifefb"
    >
      London
       
      <small
        class="svelte-1mifefb"
      >
        GMT 
        +00:00
      </small>
    </span>
     
    <svg
      class="svelte-1mifefb"
      height="16"
      viewBox="0 0 16 16"
      width="10"
    >
      <polygon
        class="svelte-1mifefb"
        points="8, 8, 16, 16, 0, 16"
        transform="rotate(0) translate(0 -4)"
        x="0"
        y="0"
      />
    </svg>
  </button>
   
  <div
    class="tz-dropdown svelte-1mifefb"
    style="animation: __svelte_276924520_0 400ms linear 0ms 1 both;"
  >
    <span
      class="sr-only svelte-1mifefb"
      id="mock-test-id-9"
    >
      Select a timezone from the list. Start typing to filter or use the arrow
      keys to navigate the list
    </span>
     
    <div
      class="input-group svelte-1mifefb"
    >
      <input
        aria-autocomplete="list"
        aria-controls="mock-test-id-10"
        aria-labelledby="mock-test-id-9"
        autocomplete="off"
        autocorrect="off"
        autofocus=""
        class="svelte-1mifefb"
        id="mock-test-id-11"
        placeholder="Search..."
        type="search"
      />
       
    </div>
     
    <ul
      aria-activedescendant="tz-london"
      aria-labelledby="mock-test-id-9"
      class="tz-groups svelte-1mifefb"
      id="mock-test-id-10"
      role="listbox"
      tabindex="-1"
    >
      <li
        aria-hidden="true"
        class="svelte-1mifefb"
        role="option"
      >
        <p
          class="svelte-1mifefb"
        >
          Australia
        </p>
      </li>
       
      <li
        aria-label="Select Sydney"
        aria-selected="false"
        class="svelte-1mifefb"
        id="tz-australiasydney"
        role="option"
        tabindex="0"
      >
        Sydney
         
        <span
          class="svelte-1mifefb"
        >
          GMT 
          +10:00
        </span>
         
      </li>
      
      <li
        aria-label="Select Melbourne"
        aria-selected="false"
        class="svelte-1mifefb"
        id="tz-australiamelbourne"
        role="option"
        tabindex="0"
      >
        Melbourne
         
        <span
          class="svelte-1mifefb"
        >
          GMT 
          +10:00
        </span>
         
      </li>
      
      
      
      <li
        aria-hidden="true"
        class="svelte-1mifefb"
        role="option"
      >
        <p
          class="svelte-1mifefb"
        >
          Europe
        </p>
      </li>
       
      <li
        aria-label="Select Berlin"
        aria-selected="false"
        class="svelte-1mifefb"
        id="tz-europeberlin"
        role="option"
        tabindex="0"
      >
        Berlin
         
        <span
          class="svelte-1mifefb"
        >
          GMT 
          +01:00
        </span>
         
      </li>
      
      <li
        aria-label="Select Dublin"
        aria-selected="false"
        class="svelte-1mifefb"
        id="tz-europedublin"
        role="option"
        tabindex="0"
      >
        Dublin
         
        <span
          class="svelte-1mifefb"
        >
          GMT 
          +00:00
        </span>
         
      </li>
      
      <li
        aria-label="Select London"
        aria-selected="false"
        class="svelte-1mifefb"
        id="tz-europelondon"
        role="option"
        tabindex="0"
      >
        London
         
        <span
          class="svelte-1mifefb"
        >
          GMT 
          +00:00
        </span>
         
      </li>
      
      <li
        aria-label="Select Lisbon"
        aria-selected="false"
        class="svelte-1mifefb"
        id="tz-europelisbon"
        role="option"
        tabindex="0"
      >
        Lisbon
         
        <span
          class="svelte-1mifefb"
        >
          GMT 
          +00:00
        </span>
         
      </li>
      
      
      
      <li
        aria-hidden="true"
        class="svelte-1mifefb"
        role="option"
      >
        <p
          class="svelte-1mifefb"
        >
          Other
        </p>
      </li>
       
      <li
        aria-label="Select UTC"
        aria-selected="false"
        class="svelte-1mifefb"
        id="tz-utc"
        role="option"
        tabindex="0"
      >
        UTC
         
        <span
          class="svelte-1mifefb"
        >
          GMT 
          +00:00
        </span>
         
      </li>
      
      
      
    </ul>
  </div>
  
</div>
`;
